File: go/indicators/johnehlers/hilberttransformer/homodynediscriminatorestimator.go
package hilberttransformer

import (
	"math"
	"sync"
)

// HomodyneDiscriminatorEstimator implements the Hilbert transformer of the
// WMA-smoothed and detrended data with the Homodyne Discriminator applied.
//
// John Ehlers, Rocket Science for Traders, Wiley, 2001, 0471405671, pp 52-77.
type HomodyneDiscriminatorEstimator struct {
	mu                               sync.RWMutex
	smoothingLength                  int
	minPeriod                        int
	maxPeriod                        int
	alphaEmaQuadratureInPhase        float64
	alphaEmaPeriod                   float64
	warmUpPeriod                     int
	smoothingLengthPlusHtLengthMin1  int
	smoothingLengthPlus2HtLengthMin2 int
	smoothingLengthPlus3HtLengthMin3 int
	smoothingLengthPlus3HtLengthMin2 int
	smoothingLengthPlus3HtLengthMin1 int
	smoothingLengthPlus3HtLength     int
	oneMinAlphaEmaQuadratureInPhase  float64
	oneMinAlphaEmaPeriod             float64
	rawValues                        []float64
	wmaFactors                       []float64
	wmaSmoothed                      []float64
	detrended                        []float64
	inPhase                          []float64
	quadrature                       []float64
	jInPhase                         []float64
	jQuadrature                      []float64
	count                            int
	smoothedInPhasePrevious          float64
	smoothedQuadraturePrevious       float64
	rePrevious                       float64
	imPrevious                       float64
	period                           float64
	isPrimed                         bool
	isWarmedUp                       bool
}

// NewHomodyneDiscriminatorEstimator returns an instnce of the estimator using supplied parameters.
func NewHomodyneDiscriminatorEstimator(p *CycleEstimatorParams) (*HomodyneDiscriminatorEstimator, error) {
	err := verifyParameters(p)
	if err != nil {
		return nil, err
	}

	length := p.SmoothingLength
	alphaQuad := p.AlphaEmaQuadratureInPhase
	alphaPeriod := p.AlphaEmaPeriod

	smoothingLengthPlusHtLengthMin1 := length + htLength - 1
	smoothingLengthPlus2HtLengthMin2 := smoothingLengthPlusHtLengthMin1 + htLength - 1
	smoothingLengthPlus3HtLengthMin3 := smoothingLengthPlus2HtLengthMin2 + htLength - 1
	smoothingLengthPlus3HtLengthMin2 := smoothingLengthPlus3HtLengthMin3 + 1
	smoothingLengthPlus3HtLengthMin1 := smoothingLengthPlus3HtLengthMin2 + 1
	smoothingLengthPlus3HtLength := smoothingLengthPlus3HtLengthMin1 + 1

	// These slices will be automatically filled with zeroes.
	wmaSmoothed := make([]float64, htLength)
	detrended := make([]float64, htLength)
	inPhase := make([]float64, htLength)
	quadrature := make([]float64, htLength)
	jInPhase := make([]float64, htLength)
	jQuadrature := make([]float64, htLength)
	rawValues := make([]float64, length)
	wmaFactors := make([]float64, length)

	fillWmaFactors(length, wmaFactors)

	return &HomodyneDiscriminatorEstimator{
		smoothingLength:                  length,
		minPeriod:                        defaultMinPeriod,
		maxPeriod:                        defaultMaxPeriod,
		alphaEmaQuadratureInPhase:        alphaQuad,
		alphaEmaPeriod:                   alphaPeriod,
		warmUpPeriod:                     max(p.WarmUpPeriod, smoothingLengthPlus3HtLength),
		smoothingLengthPlusHtLengthMin1:  smoothingLengthPlusHtLengthMin1,
		smoothingLengthPlus2HtLengthMin2: smoothingLengthPlus2HtLengthMin2,
		smoothingLengthPlus3HtLengthMin3: smoothingLengthPlus3HtLengthMin3,
		smoothingLengthPlus3HtLengthMin2: smoothingLengthPlus3HtLengthMin2,
		smoothingLengthPlus3HtLengthMin1: smoothingLengthPlus3HtLengthMin1,
		smoothingLengthPlus3HtLength:     smoothingLengthPlus3HtLength,
		oneMinAlphaEmaQuadratureInPhase:  1 - alphaQuad,
		oneMinAlphaEmaPeriod:             1 - alphaPeriod,
		rawValues:                        rawValues,
		wmaFactors:                       wmaFactors,
		wmaSmoothed:                      wmaSmoothed,
		detrended:                        detrended,
		inPhase:                          inPhase,
		quadrature:                       quadrature,
		jInPhase:                         jInPhase,
		jQuadrature:                      jQuadrature,
		period:                           defaultMinPeriod,
	}, nil
}

// SmoothingLength returns the underlying WMA smoothing length in samples.
func (s *HomodyneDiscriminatorEstimator) SmoothingLength() int {
	return s.smoothingLength
}

// MinPeriod returns the minimal cycle period.
func (s *HomodyneDiscriminatorEstimator) MinPeriod() int {
	return s.minPeriod
}

// MaxPeriod returns the maximual cycle period.
func (s *HomodyneDiscriminatorEstimator) MaxPeriod() int {
	return s.maxPeriod
}

// WarmUpPeriod returns the number of updates before the estimator
// is primed (MaxPeriod * 2 = 100).
func (s *HomodyneDiscriminatorEstimator) WarmUpPeriod() int {
	return s.warmUpPeriod
}

// AlphaEmaQuadratureInPhase returns the value of α (0 < α ≤ 1)
// used in EMA to smooth the in-phase and quadrature components.
func (s *HomodyneDiscriminatorEstimator) AlphaEmaQuadratureInPhase() float64 {
	return s.alphaEmaQuadratureInPhase
}

// AlphaEmaPeriod returns the value of α (0 < α ≤ 1)
// used in EMA to smooth the instantaneous period.
func (s *HomodyneDiscriminatorEstimator) AlphaEmaPeriod() float64 {
	return s.alphaEmaPeriod
}

// Count returns the current count value.
func (s *HomodyneDiscriminatorEstimator) Count() int {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.count
}

// Primed indicates whether an instance is primed.
func (s *HomodyneDiscriminatorEstimator) Primed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.isWarmedUp
}

// Period returns the current period value.
func (s *HomodyneDiscriminatorEstimator) Period() float64 {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.period
}

// InPhase returns the current InPhase component value.
func (s *HomodyneDiscriminatorEstimator) InPhase() float64 {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.inPhase[0]
}

// Quadrature returns the current Quadrature component value.
func (s *HomodyneDiscriminatorEstimator) Quadrature() float64 {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.quadrature[0]
}

// Detrended returns the current detrended value.
func (s *HomodyneDiscriminatorEstimator) Detrended() float64 {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.detrended[0]
}

// Smoothed returns the current WMA-smoothed value used by underlying Hilbert transformer.
//
// The linear-Weighted Moving Average has a window size of SmoothingLength.
func (s *HomodyneDiscriminatorEstimator) Smoothed() float64 {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.wmaSmoothed[0]
}

// Update updates the estimator given the next sample value.
func (s *HomodyneDiscriminatorEstimator) Update(sample float64) { //nolint:funlen, cyclop
	if math.IsNaN(sample) {
		return
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	const (
		twoPi = 2 * math.Pi
	)

	push(s.rawValues, sample)

	if s.isPrimed { //nolint:nestif
		if !s.isWarmedUp {
			s.count++
			if s.warmUpPeriod < s.count {
				s.isWarmedUp = true
			}
		}

		// The WMA is used to remove some high-frequency components before detrending the signal.
		push(s.wmaSmoothed, s.wma(s.rawValues))

		amplitudeCorrectionFactor := correctAmplitude(s.period)

		// Since we have an amplitude-corrected Hilbert transformer, and since we want to detrend
		// over its length, we simply use the Hilbert transformer itself as the detrender.
		push(s.detrended, ht(s.wmaSmoothed)*amplitudeCorrectionFactor)

		// Compute both the in-phase and quadrature components of the detrended signal.
		push(s.quadrature, ht(s.detrended)*amplitudeCorrectionFactor)
		push(s.inPhase, s.detrended[quadratureIndex])

		// Complex averaging: apply the Hilbert Transformer to both the in-phase and quadrature components.
		// This advances the phase of each component by 90°.
		push(s.jInPhase, ht(s.inPhase)*amplitudeCorrectionFactor)
		push(s.jQuadrature, ht(s.quadrature)*amplitudeCorrectionFactor)

		// Phasor addition for 3 bar averaging followed by exponential moving average smoothing.
		smoothedInPhase := s.emaQuadratureInPhase(s.inPhase[0]-s.jQuadrature[0], s.smoothedInPhasePrevious)
		smoothedQuadrature := s.emaQuadratureInPhase(s.quadrature[0]+s.jInPhase[0], s.smoothedQuadraturePrevious)

		// Homodyne discriminator. Calculate the real and imaginary components of the signal
		// of the current sample multiplied with the complex conjugate of the signal 1 sample ago.
		re := smoothedInPhase*s.smoothedInPhasePrevious + smoothedQuadrature*s.smoothedQuadraturePrevious
		im := smoothedInPhase*s.smoothedQuadraturePrevious - smoothedQuadrature*s.smoothedInPhasePrevious
		s.smoothedInPhasePrevious = smoothedInPhase
		s.smoothedQuadraturePrevious = smoothedQuadrature

		// Exponential moving average smoothing of the real and imaginary components.
		re = s.emaQuadratureInPhase(re, s.rePrevious)
		im = s.emaQuadratureInPhase(im, s.imPrevious)
		s.rePrevious = re
		s.imPrevious = im
		periodPrevious := s.period
		periodNew := twoPi / math.Atan2(im, re)

		if !math.IsNaN(periodNew) && !math.IsInf(periodNew, 0) {
			s.period = periodNew
		}

		s.period = adjustPeriod(s.period, periodPrevious)

		// Exponential moving average smoothing of the period.
		s.period = s.emaPeriod(s.period, periodPrevious)
	} else { // Not primed.
		// On (smoothingLength)-th sample we calculate the first
		// WMA smoothed value and begin with the detrender.
		s.count++
		if s.smoothingLength > s.count { // count < 4
			return
		}

		push(s.wmaSmoothed, s.wma(s.rawValues)) // count >= 4

		if s.smoothingLengthPlusHtLengthMin1 > s.count { // count < 10
			return
		}

		amplitudeCorrectionFactor := correctAmplitude(s.period) // count >= 10
		push(s.detrended, ht(s.wmaSmoothed)*amplitudeCorrectionFactor)

		if s.smoothingLengthPlus2HtLengthMin2 > s.count { // count < 16
			return
		}

		push(s.quadrature, ht(s.detrended)*amplitudeCorrectionFactor) // count >= 16
		push(s.inPhase, s.detrended[quadratureIndex])

		if s.smoothingLengthPlus3HtLengthMin3 > s.count { // count < 22
			return
		}

		push(s.jInPhase, ht(s.inPhase)*amplitudeCorrectionFactor) // count >= 22
		push(s.jQuadrature, ht(s.quadrature)*amplitudeCorrectionFactor)

		if s.smoothingLengthPlus3HtLengthMin3 == s.count { // count == 22
			s.smoothedInPhasePrevious = s.inPhase[0] - s.jQuadrature[0]
			s.smoothedQuadraturePrevious = s.quadrature[0] + s.jInPhase[0]

			return
		}

		smoothedInPhase := s.emaQuadratureInPhase(s.inPhase[0]-s.jQuadrature[0], s.smoothedInPhasePrevious) // count >= 23
		smoothedQuadrature := s.emaQuadratureInPhase(s.quadrature[0]+s.jInPhase[0], s.smoothedQuadraturePrevious)

		re := smoothedInPhase*s.smoothedInPhasePrevious + smoothedQuadrature*s.smoothedQuadraturePrevious
		im := smoothedInPhase*s.smoothedQuadraturePrevious - smoothedQuadrature*s.smoothedInPhasePrevious
		s.smoothedInPhasePrevious = smoothedInPhase
		s.smoothedQuadraturePrevious = smoothedQuadrature

		if s.smoothingLengthPlus3HtLengthMin2 == s.count { // count == 23
			s.rePrevious = re
			s.imPrevious = im

			return
		}

		re = s.emaQuadratureInPhase(re, s.rePrevious) // count >= 24
		im = s.emaQuadratureInPhase(im, s.imPrevious)
		s.rePrevious = re
		s.imPrevious = im
		periodPrevious := s.period

		periodNew := twoPi / math.Atan2(im, re)
		if !math.IsNaN(periodNew) && !math.IsInf(periodNew, 0) {
			s.period = periodNew
		}

		s.period = adjustPeriod(s.period, periodPrevious)

		if s.smoothingLengthPlus3HtLengthMin1 < s.count { // count > 24
			s.period = s.emaPeriod(s.period, periodPrevious)
			s.isPrimed = true
		}
	}
}

func (s *HomodyneDiscriminatorEstimator) wma(array []float64) float64 {
	value := 0.
	for i := range s.smoothingLength {
		value += s.wmaFactors[i] * array[i]
	}

	return value
}

func (s *HomodyneDiscriminatorEstimator) emaQuadratureInPhase(value, valuePrevious float64) float64 {
	return s.alphaEmaQuadratureInPhase*value + s.oneMinAlphaEmaQuadratureInPhase*valuePrevious
}

func (s *HomodyneDiscriminatorEstimator) emaPeriod(value, valuePrevious float64) float64 {
	return s.alphaEmaPeriod*value + s.oneMinAlphaEmaPeriod*valuePrevious
}

File: py/indicators/john_ehlers/hilbert_transformer/homodyne_discriminator_estimator.py
"""Homodyne Discriminator cycle estimator implementation."""
import math

from .cycle_estimator import CycleEstimator
from .cycle_estimator_params import CycleEstimatorParams
from .estimator import (
    DEFAULT_MIN_PERIOD, DEFAULT_MAX_PERIOD, HT_LENGTH, QUADRATURE_INDEX,
    push, ht, correct_amplitude, adjust_period, fill_wma_factors, verify_parameters,
)


class HomodyneDiscriminatorEstimator(CycleEstimator):
    """Implements the Hilbert transformer with the Homodyne Discriminator.

    John Ehlers, Rocket Science for Traders, Wiley, 2001, 0471405671, pp 52-77.
    """

    def __init__(self, p: CycleEstimatorParams) -> None:
        verify_parameters(p)

        length = p.smoothing_length
        alpha_quad = p.alpha_ema_quadrature_in_phase
        alpha_period = p.alpha_ema_period

        sl_ht1 = length + HT_LENGTH - 1
        sl_ht2 = sl_ht1 + HT_LENGTH - 1
        sl_ht3 = sl_ht2 + HT_LENGTH - 1

        self._smoothing_length = length
        self._min_period = DEFAULT_MIN_PERIOD
        self._max_period = DEFAULT_MAX_PERIOD
        self._alpha_ema_quadrature_in_phase = alpha_quad
        self._alpha_ema_period = alpha_period
        self._warm_up_period = max(p.warm_up_period, sl_ht3 + 3)
        self._sl_ht1 = sl_ht1
        self._sl_ht2 = sl_ht2
        self._sl_ht3 = sl_ht3
        self._sl_ht3_p1 = sl_ht3 + 1
        self._sl_ht3_p2 = sl_ht3 + 2
        self._sl_ht3_p3 = sl_ht3 + 3
        self._one_min_alpha_quad = 1.0 - alpha_quad
        self._one_min_alpha_period = 1.0 - alpha_period

        self._raw_values = [0.0] * length
        self._wma_factors = [0.0] * length
        self._wma_smoothed = [0.0] * HT_LENGTH
        self._detrended = [0.0] * HT_LENGTH
        self._in_phase_arr = [0.0] * HT_LENGTH
        self._quadrature_arr = [0.0] * HT_LENGTH
        self._j_in_phase = [0.0] * HT_LENGTH
        self._j_quadrature = [0.0] * HT_LENGTH

        fill_wma_factors(length, self._wma_factors)

        self._count = 0
        self._smoothed_in_phase_prev = 0.0
        self._smoothed_quadrature_prev = 0.0
        self._re_prev = 0.0
        self._im_prev = 0.0
        self._period_val = float(DEFAULT_MIN_PERIOD)
        self._is_primed = False
        self._is_warmed_up = False

    def smoothing_length(self) -> int:
        return self._smoothing_length

    def smoothed(self) -> float:
        return self._wma_smoothed[0]

    def detrended(self) -> float:
        return self._detrended[0]

    def quadrature(self) -> float:
        return self._quadrature_arr[0]

    def in_phase(self) -> float:
        return self._in_phase_arr[0]

    def period(self) -> float:
        return self._period_val

    def count(self) -> int:
        return self._count

    def primed(self) -> bool:
        return self._is_warmed_up

    def min_period(self) -> int:
        return self._min_period

    def max_period(self) -> int:
        return self._max_period

    def alpha_ema_quadrature_in_phase(self) -> float:
        return self._alpha_ema_quadrature_in_phase

    def alpha_ema_period(self) -> float:
        return self._alpha_ema_period

    def warm_up_period(self) -> int:
        return self._warm_up_period

    def update(self, sample: float) -> None:
        if math.isnan(sample):
            return

        two_pi = 2.0 * math.pi

        push(self._raw_values, sample)

        if self._is_primed:
            if not self._is_warmed_up:
                self._count += 1
                if self._warm_up_period < self._count:
                    self._is_warmed_up = True

            push(self._wma_smoothed, self._wma(self._raw_values))
            acf = correct_amplitude(self._period_val)

            push(self._detrended, ht(self._wma_smoothed) * acf)
            push(self._quadrature_arr, ht(self._detrended) * acf)
            push(self._in_phase_arr, self._detrended[QUADRATURE_INDEX])
            push(self._j_in_phase, ht(self._in_phase_arr) * acf)
            push(self._j_quadrature, ht(self._quadrature_arr) * acf)

            smoothed_ip = self._ema_quad(
                self._in_phase_arr[0] - self._j_quadrature[0],
                self._smoothed_in_phase_prev)
            smoothed_q = self._ema_quad(
                self._quadrature_arr[0] + self._j_in_phase[0],
                self._smoothed_quadrature_prev)

            re = smoothed_ip * self._smoothed_in_phase_prev + \
                smoothed_q * self._smoothed_quadrature_prev
            im = smoothed_ip * self._smoothed_quadrature_prev - \
                smoothed_q * self._smoothed_in_phase_prev
            self._smoothed_in_phase_prev = smoothed_ip
            self._smoothed_quadrature_prev = smoothed_q

            re = self._ema_quad(re, self._re_prev)
            im = self._ema_quad(im, self._im_prev)
            self._re_prev = re
            self._im_prev = im
            period_prev = self._period_val
            atan_val = math.atan2(im, re)
            if atan_val == 0.0:
                period_new = math.inf
            else:
                period_new = two_pi / atan_val

            if not math.isnan(period_new) and not math.isinf(period_new):
                self._period_val = period_new

            self._period_val = adjust_period(self._period_val, period_prev)
            self._period_val = self._ema_period(self._period_val, period_prev)
        else:
            self._count += 1
            if self._smoothing_length > self._count:
                return

            push(self._wma_smoothed, self._wma(self._raw_values))

            if self._sl_ht1 > self._count:
                return

            acf = correct_amplitude(self._period_val)
            push(self._detrended, ht(self._wma_smoothed) * acf)

            if self._sl_ht2 > self._count:
                return

            push(self._quadrature_arr, ht(self._detrended) * acf)
            push(self._in_phase_arr, self._detrended[QUADRATURE_INDEX])

            if self._sl_ht3 > self._count:
                return

            push(self._j_in_phase, ht(self._in_phase_arr) * acf)
            push(self._j_quadrature, ht(self._quadrature_arr) * acf)

            if self._sl_ht3 == self._count:
                self._smoothed_in_phase_prev = \
                    self._in_phase_arr[0] - self._j_quadrature[0]
                self._smoothed_quadrature_prev = \
                    self._quadrature_arr[0] + self._j_in_phase[0]
                return

            smoothed_ip = self._ema_quad(
                self._in_phase_arr[0] - self._j_quadrature[0],
                self._smoothed_in_phase_prev)
            smoothed_q = self._ema_quad(
                self._quadrature_arr[0] + self._j_in_phase[0],
                self._smoothed_quadrature_prev)

            re = smoothed_ip * self._smoothed_in_phase_prev + \
                smoothed_q * self._smoothed_quadrature_prev
            im = smoothed_ip * self._smoothed_quadrature_prev - \
                smoothed_q * self._smoothed_in_phase_prev
            self._smoothed_in_phase_prev = smoothed_ip
            self._smoothed_quadrature_prev = smoothed_q

            if self._sl_ht3_p1 == self._count:
                self._re_prev = re
                self._im_prev = im
                return

            re = self._ema_quad(re, self._re_prev)
            im = self._ema_quad(im, self._im_prev)
            self._re_prev = re
            self._im_prev = im
            period_prev = self._period_val

            atan_val = math.atan2(im, re)
            if atan_val != 0.0:
                period_new = two_pi / atan_val
            else:
                period_new = math.inf
            if not math.isnan(period_new) and not math.isinf(period_new):
                self._period_val = period_new

            self._period_val = adjust_period(self._period_val, period_prev)

            if self._sl_ht3_p2 < self._count:
                self._period_val = self._ema_period(self._period_val, period_prev)
                self._is_primed = True

    def _wma(self, array: list[float]) -> float:
        value = 0.0
        for i in range(self._smoothing_length):
            value += self._wma_factors[i] * array[i]
        return value

    def _ema_quad(self, value: float, value_prev: float) -> float:
        return self._alpha_ema_quadrature_in_phase * value + \
            self._one_min_alpha_quad * value_prev

    def _ema_period(self, value: float, value_prev: float) -> float:
        return self._alpha_ema_period * value + \
            self._one_min_alpha_period * value_prev
